nightfall-client/src/services/keys.mjs

```diff
@@ -16,57 +31,55 @@ export function compressPublicKey(publicKey) {
   return new GN(edwardsCompress([publicKey[0].bigInt, publicKey[1].bigInt]));
 }
 
-async function generateASK(length) {
-  const ask = new GN((await rand(length)).bigInt % BN128_GROUP_ORDER);
-  return ask;
-}
-
-async function generateNSK(length) {
-  const nsk = new GN((await rand(length)).bigInt % BN128_GROUP_ORDER);
-  return nsk;
+// path structure is m / purpose' / coin_type' / account' / change / address_index
+// the path we use is m/44'/60'/account'/0/address_index
+// address will change incrementally for every new set of keys to be created from the same seed
+// address_index will define if the key is ask or nsk
+// path for ask is m/44'/60'/account'/0/0
+// path for nsk is m/44'/60'/account'/0/1
+async function generateASKAndNSK(seed, path) {
+  const ask = generalise(generatePrivateKey(seed, `${path}/0`), 'bigInt');
+  const nsk = generalise(generatePrivateKey(seed, `${path}/1`), 'bigInt');
+  return { ask, nsk };
 }
 
+// Calculate ivk from ask and nsk such as ivk = MiMC(ask, nsk)
 function calculateIVK(ask, nsk) {
   return new GN(mimcHash([ask.bigInt, nsk.bigInt]));
 }
 
+// Calculate pkd from ivk such as pkd = ivk.G
 export function calculatePkd(ivk) {
   const pkd = calculatePublicKey(ivk);
   const compressedPkd = compressPublicKey(pkd);
   return { pkd, compressedPkd };
 }
 
-// function calculatePkdFromAskNsk(ask, nsk) {
-//   const ivk = calculateIVK(ask, nsk);
-//   return calculatePkd(ivk);
-// }
-
-export function calculateIvkPkdfromAskNsk(ask, nsk) {
-  const ivk = calculateIVK(ask, nsk);
-  const { pkd, compressedPkd } = calculatePkd(ivk);
-  return { ivk, pkd, compressedPkd };
+// function to generate all the required keys deterministically from a random mnemonic
+// Use mnemonic to generate seed which will then be used to generate sets of ask and nsk based on different account numbers
+export async function generateKeys(mnemonic, path) {
```

Comment:
I was hoping to receive a number of zkpKeys. Can we add this functionality in a future revision? If i understand our conversation, you would concatenate the seed obtained in `generateHexSeed` with a sequence number in order to generate multiple keys. Is this correct?

Reply:
the path will change to generate new keys each time. The seed and the mnemonic for technical purposes can be considered synonymous. According to the BIP39 standard, there seem to be a bijective relationship between mnemonic and seed. Each mnemonic has a unique seed. We use this and a different path to generate different keys each time. 

Do you want to generate multiple accounts (different pairs of zkp keys) at once ?

Reply:
Ok, i understand. Based on the first question i posted, I understand how to generate several keys. Your plan is to simply pass a different index based on the same mnemonic and a different set of keys will be generated. I am ok with this, so no change necessary.